src/transport/SecureSessionMgr.h

```diff
@@ -92,13 +94,19 @@ class EncryptedPacketBufferHandle final : private System::PacketBufferHandle
     CHIP_ERROR InsertPacketHeader(const PacketHeader & aPacketHeader) { return aPacketHeader.EncodeBeforeData(*this); }
 #endif // CHIP_ENABLE_TEST_ENCRYPTED_BUFFER_API
 
+    // Making operator= available to tests by making it public.
```

Comment:
Why is this needed? this looks very suspect especially with the `#ifdef`

Reply:
I think making this public should be ok.

Reply:
@yufengwangca - why where the constructors and assignment operators made private?

It seems we have this pattern of 'private constructor' and friend class which causes coupling isssues.

If we need to avoid casting things by mistake, how about providing a

```cpp
static EncryptedPacketBufferHandle MarkPacketEncrypted(PacketBufferHandle&& buffer);
```

which would make it clear that we have a packet that is encrypted and code was written on purpose to do this. I would completely drop the 'friend class'.

Reply:
I believe this was to ensure that Encrypted packet can only be created by SecureSessionMgr. I.e. to guarantee that the packet being retransmitted is correctly encrypted, and retransmit logic cannot accidentally send an unencrypted packet.

This, condition may be relaxed now, since for PASE/CASE session establishment messages, the retransmitted packet will not be encrypted.

I'll move it out of the ifdef condition.

Reply:
I believe we should move away from coupling using friend classes.

Reply:
Just to give the history here: the idea was that becuase EncryptedPacketBufferHandle is transmitted without encryption it gives a way to transmit things without encryption in general if you can control the contents of the handle.  Which is not at all desirable. So all the ways to create an EncryptedPacketBufferHandle with random data that is not encrypted were marked private, and only the class that does encryption was allowed to do those things.

The reason this is coming up here is that in the test we _want_ to inject data of our choosing into an EncryptedPacketBufferHandle.  For one thing, that's the only way to test malformed data.  But ideally we would keep the API surface for doing this as small as possible, ideally test-only, and not expose it in general.

@andy31415 Any suggestions on how you would like to see this set of requirements addressed?

Reply:
I believe having a 'MarkPacketBufferEncrypted' is acceptable as public and we should not have a friend class at all.

The class provides compile-type safety saying "you need an encrypted buffer here" and the developer has to explicitly mark buffers as encrypted.

Marking things private means that instead of `MarkEncrypted` a "hacky developer" would need to either change chip SDK sligtly (open source after all) or use some 'reinterpret cast' to change internal pointers. Either way, I do not want to fight developers and would rather have a clean API that is hard to misuse. I believe a 'mark is encrypted' API is clear enough and if misused it is so intentionally.